README.md
# dbanon

[![Release Status](https://github.com/mdshack/dbanon/actions/workflows/release-please.yaml/badge.svg)](https://github.com/mdshack/dbanon/actions) [![codecov](https://codecov.io/gh/mpchadwick/dbanon/branch/master/graph/badge.svg)](https://codecov.io/gh/mdshack/dbanon)

A run-anywhere, dependency-less database anonymizer.

## Installation

Download [the latest release from GitHub](https://github.com/mdshack/dbanon/releases).

## Usage

`dbanon` reads from `stdin` and writes to `stdout`.

```
mysqldump mydb | dbanon -config=myconfig.yml | gzip > mydb.sql.gz
```

The `-config` flag can use bundled configurations or point to the path of a custom configuration file. 

### Rules

#### Static

These are basic faking rules that don't accept any arguments or know/care about the current value of a cell.

##### `first_name`
##### `last_name`
##### `full_name`
##### `company_name`
##### `username`
##### `password`
##### `ipv4`
##### `url`
##### `linkedin_url`
##### `md5`
##### `uuid`
##### `json`
##### `query_params`
##### `datetime`
##### `state`
##### `city`
##### `postcode`
##### `street`
##### `country_code`
##### `money`
##### `money_decimal`
##### `email`
##### `telephone`

#### Dynamic

These are more advanced rules that accept arguments and can be impacted by the current value of the cell.

##### `dynamic.email`
**arguments**
- `ignoreEmail` - Any emails with this suffix will not be faked.

**example**
```
rule = dyanmic.email(@test.com)
current cell value = blam@test.com
after dbanon = blam@test.com

rule = dyanmic.email(@test.com)
current cell value = test@blam.com
after dbanon = random@fake.email
```

### Configuration

#### Magento 2

`dbanon` bundles a [default Magento 2 configuration file](etc/magento2.yml). However you almost certainly won't use it directly.

At minimum, you'll first need to run the `map-eav` subcommand. This translates EAV attribute codes to their respective attribute ids.

You must feed it a `mysqldump` of `eav_entity_type` and `eav_attribute` (in that order).

```
mysqldump mydb eav_entity_type eav_attribute | dbanon -config=magento2 map-eav > ~/magento2-mapped.yml
```

`map-eav` will replace the attribute codes in the config file with attribute ids and print an updated config to `stdout`.

Next you'd run `dbanon` with the config generated by `map-eav`.

```
mysqldump mydb | dbanon -config=~/magento2-mapped.yml | gzip > mydb.sql.gz
```


Most Magento 2 databases, however, will have additional data that needs to be anonymized beyond the default bundled file. 

For this you'll first want to create a new configuration file based off the bundled configuration. Instructions on customizing the configuration file are included in the "Custom Configuration" section.


#### Custom Configuration

Specify the path to your config file via the `-config` flag

```
mysqldump mydb | dbanon -config=myconfig.yml | gzip > mydb.sql.gz
```

See [the `etc` directory](etc/) for examples.

Columns are specified as key / value pairs. The value string winds up getting passed to [this function](https://github.com/mdshack/dbanon/blob/ade634a10bc282c06fecef115afbdd6661a94277/src/provider.go#L36), which gets random values from [`dmgk/faker`](https://github.com/dmgk/faker).

It is also possible to pass direct Faker function calls for [supported "raw providers"](https://github.com/mdshack/dbanon/blob/ade634a10bc282c06fecef115afbdd6661a94277/src/provider.go#L13-L17)

## Logging

`dbanon` records messages about anything notable (e.g. invalid configuration) to the file `dbanon.log` in the directory from which you run it.

**`-log-file`**

The `-log-file` flag can be used to have `dbanon` log to a different location.

```
mysqldump mydb | dbanon -config=myconfig.yml -log-file=var/dbanon.log
```

**`-log-level`**

The `-log-level` flag can be used to control the verbosity of logs. Supported values can be found [here](https://github.com/sirupsen/logrus/blob/d131c24e23baaa812461202af6d7cfa388e2d292/logrus.go#L25-L45).

```
mysqldump mydb | dbanon -config=myconfig.yml -log-level=debug | gzip > mydb.sql.gz
```

The default log level is `info`.

**`-silent`**

Logging can be disabled entirely by passing the `-silent` flag to `dbanon`

```
mysqldump mydb | dbanon -config=myconfig.yml -silent | gzip > mydb.sql.gz
```

## Profiling

`dbanon` will generate a CPU profile to the file `dbanon.prof` when passed the `-profile` flag.

```
mysqldump mydb | dbanon -profile -config=myconfig.yml >/dev/null
```

## Limitations

- Currently only supports MySQL

## Updating

`dbanon` will self-update when passed the `-update` flag

```
dbanon -update
```

## Development

### Publishing releases

Releases are published with [GoReleaser](https://goreleaser.com/).

- Create and push a tag (e.g. `git tag -a v0.6.2 -m "v0.6.2 && git push origin v0.6.2`)
- Build the `go-bindata` (e.g. `go-bindata -pkg bindata -o bindata/bindata.go etc/*`)
- Run `goreleaser` (e.g. `goreleaser release --rm-dist`)

May need to create a new GitHub token if the current one expires

https://goreleaser.com/scm/github/


CHANGELOG.md
# Changelog

## [1.2.0](https://github.com/mdshack/dbanon/compare/v1.1.3...v1.2.0) (2023-12-14)


### Features

* migrate to vitess.io/vitess/go/vt/sqlparser ([65f0bd6](https://github.com/mdshack/dbanon/commit/65f0bd6dd22adb5970a58f737d883c2ab8e66f68))

## [1.1.3](https://github.com/mdshack/dbanon/compare/v1.1.2...v1.1.3) (2023-09-23)


### Bug Fixes

* drop bindata ([3c31208](https://github.com/mdshack/dbanon/commit/3c31208f98a4b3a6d2216c626ab2b4b3c78dc410))

## [1.1.2](https://github.com/mdshack/dbanon/compare/v1.1.1...v1.1.2) (2023-09-23)


### Bug Fixes

* release ([017db6d](https://github.com/mdshack/dbanon/commit/017db6d4bff09d3007148f9b10ebf6438bc7643c))

## [1.1.1](https://github.com/mdshack/dbanon/compare/v1.1.0...v1.1.1) (2023-09-23)


### Bug Fixes

* release id ([5fec831](https://github.com/mdshack/dbanon/commit/5fec83106ec2f25180714897adab063e7a079b79))

## [1.1.0](https://github.com/mdshack/dbanon/compare/v1.0.1...v1.1.0) (2023-09-23)


### Features

* gorelease ([dc816ba](https://github.com/mdshack/dbanon/commit/dc816ba63f2c4ec2d6306932e39976f57f88d31c))
* release jobs ([ba42a66](https://github.com/mdshack/dbanon/commit/ba42a66d18d805365de24e2b203626e93b353c05))

## [1.0.1](https://github.com/mdshack/dbanon/compare/v1.0.0...v1.0.1) (2023-09-23)


### Bug Fixes

* add vendor ([2543d15](https://github.com/mdshack/dbanon/commit/2543d15276ee1931ebad7c7cbaba6fd36f407a09))

## 1.0.0 (2023-09-23)


### Features

* add json & query param providers ([456fd65](https://github.com/mdshack/dbanon/commit/456fd65a822e3b3208249759ad3888f684053717))
* current value aware fakes, dynamic args, etc ([c0ae67f](https://github.com/mdshack/dbanon/commit/c0ae67f20c0a5962671d6243bb3f7759d460d5b6))


src/provider_test.go
package dbanon

import (
	"github.com/sirupsen/logrus/hooks/test"
	"strconv"
	"testing"
	"time"
)

func TestGetForEtcCases(t *testing.T) {
	testLogger, hook := test.NewNullLogger()
	SetLogger(testLogger)

	val := "test"

	provider := NewProvider()
	_ = provider.Get("faker.Whoops1", &val)
	if hook.LastEntry().Message != "faker.Whoops1 does not match any known type" {
		t.Errorf("Unsupported provider not handled correctly")
	}

	_ = provider.Get("faker.Number().Whoops2()", &val)
	if hook.LastEntry().Message != "faker.Number().Whoops2() does not match any known type" {
		t.Errorf("Unsupported method not handled correctly")
	}

	to := time.Now()
	from := to.AddDate(-40, 0, 0)
	r11a := provider.Get("datetime", &val)
	r11Time, _ := time.Parse("2006-01-02 15:04:05", r11a)
	if r11Time.Before(from) || r11Time.After(to) {
		t.Errorf("%v not in expected range [%v, %v]", r11Time, from, to)
	}
}

func TestGetForLengthBasedOptions(t *testing.T) {
	provider := NewProvider()
	tests := map[string]struct {
		input  string
		wantGt int
		wantLt int
		isStr  bool
	}{
		"md5":                            {input: "md5", wantGt: 31, wantLt: 33, isStr: true},
	}

	val := "test"

	for name, tc := range tests {
		t.Run(name, func(t *testing.T) {
			got := provider.Get(tc.input, &val)
			var compare int
			if tc.isStr {
				compare = len(got)
			} else {
				compare, _ = strconv.Atoi(got)
			}

			if compare < tc.wantGt || compare > tc.wantLt {
				t.Errorf("Expected %v to be greater than %v and less than %v", compare, tc.wantGt, tc.wantLt)
			}
		})
	}
}

func TestDynamicEmailProviderIgnoresEmail(t *testing.T) {
	provider := NewProvider()

	current := "blam@test.com"
	value := provider.Get("dynamic.email(@test.com)", &current)
	if current != value {
		t.Errorf("new:%s does not equal original:%s when a fake was not expected", value, current)
	}
}

func TestDynamicEmailProviderFakesEmail(t *testing.T) {
	provider := NewProvider()
	
	current := "blam@blam.com"
	value := provider.Get("dynamic.email(@test.com)", &current)
	if current == value {
		t.Errorf("new:%s equals original:%s when a fake was expected", value, current)
	}
}

func TestDynamicEmailProviderHandlesNoArgs(t *testing.T) {
	provider := NewProvider()
	
	current := "blam@blam.com"
	faked1 := provider.Get("dynamic.email()", &current)
	faked2 := provider.Get("dynamic.email", &current)
	if faked1 == current || faked2 == current {
		t.Error("unhandled empty args")
	}
}


src/processor_test.go
package dbanon

import (
	"bufio"
	"os"
	"reflect"
	"strings"
	"testing"
)

type TestProvider struct{}

func NewTestProvider() *TestProvider {
	p := &TestProvider{}

	return p
}

func (p TestProvider) Get(s string, currentValue *string) string {
	return s
}

func BenchmarkProcessLine(b *testing.B) {
	pwd, _ := os.Getwd()
	config, _ := NewConfig(pwd + "/../benchmark/laravel1.yml")
	provider := NewProvider()
	mode := "anonymize"
	eav := NewEav(config)
	processor := NewLineProcessor(mode, config, provider, eav)

	for n := 0; n < b.N; n++ {
		f, _ := os.Open(pwd + "/../benchmark/laravel1.sql")
		reader := bufio.NewReader(f)
		for {
			text, err := reader.ReadString('\n')
			_ = processor.ProcessLine(text)
			if err != nil {
				break
			}
		}

	}
}

func TestProcessLine(t *testing.T) {
	config, _ := NewConfig("magento2")
	provider := NewTestProvider()
	mode := "anonymize"
	eav := NewEav(config)
	processor := NewLineProcessor(mode, config, provider, eav)

	r1 := processor.ProcessLine("foobar")
	if r1 != "foobar" {
		t.Errorf("Got %s wanted foobar", r1)
	}

	processor.ProcessLine("CREATE TABLE `admin_user` (")
	processor.ProcessLine("  `firstname` varchar(32) DEFAULT NULL COMMENT 'User First Name'")
	processor.ProcessLine(") ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=utf8 COMMENT='Admin User Table'")
	processor.ProcessLine("/*!40101 SET character_set_client = @saved_cs_client */;")

	r2 := processor.ProcessLine("INSERT INTO `admin_user` (`firstname`) VALUES ('bob'),('joe'),('steve');")
	if strings.Contains(r2, "bob") {
		t.Error("Got bob wanted no bob")
	}

	r2b := processor.ProcessLine("INSERT INTO `admin_user` VALUES ('joe');")
	if strings.Contains(r2b, "joe") {
		t.Error("Got joe wanted no joe")
	}

	processor.ProcessLine("CREATE TABLE `admin_user` (")
	processor.ProcessLine("  `user_id` int(10) unsigned NOT NULL AUTO_INCREMENT COMMENT 'User ID'")
	processor.ProcessLine(") ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=utf8 COMMENT='Admin User Table'")
	processor.ProcessLine("/*!40101 SET character_set_client = @saved_cs_client */;")

	r3 := processor.ProcessLine("INSERT INTO `admin_user` (`user_id`) VALUES (1337);")
	if !strings.Contains(r3, "1337") {
		t.Error("Got no 1337 wanted 1337")
	}

	// todod: Not really expecting to receive input like this, but there are several reports
	// that the sqlparser.Parse is choking. Update this when we receive an example on
	// one of these GitHub issues, or on any other new issue:
	// https://github.com/mdshack/dbanon/issues/18
	// https://github.com/mdshack/dbanon/issues/17
	r3aInput := "INSERT INTO this is not valid sql"
	r3aResult := processor.ProcessLine(r3aInput)
	if r3aResult != r3aInput {
		t.Errorf("Got %s wanted empty %s", r3aResult, r3aInput)
	}

	for _, e := range processor.Config.Eav {
		if e.Name == "customer" {
			e.Attributes["1"] = "firstname"
		}
	}

	processor.ProcessLine("CREATE TABLE `customer_entity_varchar` (")
	processor.ProcessLine("  `attribute_id` smallint(5) unsigned NOT NULL DEFAULT '0' COMMENT 'Attribute Id',")
	processor.ProcessLine("  `value` varchar(255) DEFAULT NULL COMMENT 'Value'")
	processor.ProcessLine(") ENGINE=InnoDB DEFAULT CHARSET=utf8 COMMENT='Customer Entity Varchar';")
	processor.ProcessLine("/*!40101 SET character_set_client = @saved_cs_client */;")

	r4 := processor.ProcessLine("INSERT INTO `customer_entity_varchar` (`attribute_id`, `value`) VALUES (1, 'bob');")
	if strings.Contains(r4, "bob") {
		t.Error("Got bob wanted no bob")
	}

	r4b := processor.ProcessLine("INSERT INTO `customer_entity_varchar` VALUES (1, 'joe');")
	if strings.Contains(r4b, "joe") {
		t.Error("Got joe wanted no joe")
	}

	r4c := processor.ProcessLine("INSERT INTO `customer_entity_varchar` VALUES (2, 'jane');")
	if !strings.Contains(r4c, "jane") {
		t.Error("Got no jane wanted jane")
	}

	processor.ProcessLine("CREATE TABLE `migrations` (")
	processor.ProcessLine("  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,")
	processor.ProcessLine("  `migration` varchar(255) COLLATE utf8mb4_unicode_ci NOT NULL,")
	processor.ProcessLine("  `batch` int(11) NOT NULL,")
	processor.ProcessLine("  PRIMARY KEY (`id`)")
	processor.ProcessLine(") ENGINE=InnoDB AUTO_INCREMENT=4 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;")
	processor.ProcessLine("/*!40101 SET character_set_client = @saved_cs_client */;")

	r5before := "INSERT INTO `migrations` VALUES (1,'2014_10_12_000000_create_users_table',1),(2,'2014_10_12_100000_create_password_resets_table',1),(3,'2019_08_19_000000_create_failed_jobs_table',1)"
	r5after := processor.ProcessLine(r5before)
	if !reflect.DeepEqual(r5before, r5after) {
		t.Errorf("Expected %v to equal %v", r5before, r5after)
	}

}

func TestEavProcessLine(t *testing.T) {
	config, _ := NewConfig("magento2")
	provider := NewTestProvider()
	mode := "map-eav"
	eav := NewEav(config)
	processor := NewLineProcessor(mode, config, provider, eav)

	processor.ProcessLine("CREATE TABLE `eav_entity_type` (")
	processor.ProcessLine("  `entity_type_id` smallint(5) unsigned NOT NULL AUTO_INCREMENT COMMENT 'Entity Type ID',")
	processor.ProcessLine("  `entity_type_code` varchar(50) NOT NULL COMMENT 'Entity Type Code'")
	processor.ProcessLine(") ENGINE=InnoDB AUTO_INCREMENT=10 DEFAULT CHARSET=utf8 COMMENT='Eav Entity Type';")
	processor.ProcessLine("/*!40101 SET character_set_client = @saved_cs_client */;")
	processor.ProcessLine("INSERT INTO `eav_entity_type` (`entity_type_id`, `entity_type_code`) VALUES (1, 'customer');")

	processor.ProcessLine("CREATE TABLE `eav_attribute` (")
	processor.ProcessLine("  `attribute_id` smallint(5) unsigned NOT NULL AUTO_INCREMENT COMMENT 'Attribute ID',")
	processor.ProcessLine("  `entity_type_id` smallint(5) unsigned NOT NULL DEFAULT '0' COMMENT 'Entity Type ID',")
	processor.ProcessLine("  `attribute_code` varchar(255) NOT NULL COMMENT 'Attribute Code'")
	processor.ProcessLine(") ENGINE=InnoDB AUTO_INCREMENT=180 DEFAULT CHARSET=utf8 COMMENT='Eav Attribute';")
	processor.ProcessLine("/*!40101 SET character_set_client = @saved_cs_client */;")
	processor.ProcessLine("INSERT INTO `eav_attribute` (`attribute_id`, `entity_type_id`, `attribute_code`) VALUES (1, 1, 'firstname');")
	processor.ProcessLine("INSERT INTO `eav_attribute` VALUES (2, 1, 'lastname');")
	r1 := false
	r2 := false
	for _, eavConfig := range eav.Config.Eav {
		for k, v := range eavConfig.Attributes {
			if k == "1" && v == "firstname" {
				r1 = true
			}
			if k == "2" && v == "lastname" {
				r2 = true
			}
		}
	}

	if !r1 {
		t.Errorf("Got false wanted true")
	}

	if !r2 {
		t.Errorf("Got false wanted true")
	}

}


main.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"github.com/blang/semver"
	"github.com/mdshack/dbanon/src"
	"github.com/rhysd/go-github-selfupdate/selfupdate"
	"github.com/sirupsen/logrus"
	"io/ioutil"
	"log"
	"os"
	"runtime/pprof"
)

var version string

const slug = "mpchadwick/dbanon"

func selfUpdate() error {
	previous := semver.MustParse(version)
	latest, err := selfupdate.UpdateSelf(previous, slug)
	if err != nil {
		return err
	}

	if previous.Equals(latest.Version) {
		fmt.Println("Current binary is the latest version", version)
	} else {
		fmt.Println("Update successfully done to version", latest.Version)
		fmt.Println("Release note:\n", latest.ReleaseNotes)
	}
	return nil
}

func main() {
	flag.CommandLine = flag.NewFlagSet(os.Args[0], flag.ExitOnError)

	requested := flag.String("config", "", "Configuration to use. magento2 is included out-of-box. Alternately, supply path to file")
	update := flag.Bool("update", false, "Auto update dbanon to the newest version")
	ver := flag.Bool("version", false, "Get current version")
	silent := flag.Bool("silent", false, "Disable all logging")
	logFile := flag.String("log-file", "", "File to write logs to")
	logLevel := flag.String("log-level", "", "Specify desired log level")
	profile := flag.Bool("profile", false, "Generate a profile")

	flag.Parse()

	if *profile {
		profF, _ := os.Create("dbanon.prof")
		pprof.StartCPUProfile(profF)
		defer pprof.StopCPUProfile()
	}

	if *ver {
		fmt.Println(version)
		os.Exit(0)
	}

	if *update {
		if err := selfUpdate(); err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}
		os.Exit(0)
	}

	dbanonLogger := dbanon.GetLogger()
	if !*silent {
		f := "dbanon.log"
		if *logFile != "" {
			f = *logFile
		}
		file, _ := os.OpenFile(f, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
		dbanonLogger.SetOutput(file)
	} else {
		dbanonLogger.SetOutput(ioutil.Discard)
	}

	if *logLevel != "" {
		level, err := logrus.ParseLevel(*logLevel)
		if err != nil {
			dbanonLogger.Error(err)
		} else {
			dbanonLogger.SetLevel(level)
		}
	}

	config, err := dbanon.NewConfig(*requested)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}

	// sqlparser can be noisy
	// https://github.com/xwb1989/sqlparser/blob/120387863bf27d04bc07db8015110a6e96d0146c/ast.go#L52
	// We don't want to hear about it
	log.SetOutput(ioutil.Discard)
	reader := bufio.NewReader(os.Stdin)

	args := flag.Args()
	mode := "anonymize"
	if len(args) > 0 && args[0] == "map-eav" {
		mode = "map-eav"
	}

	provider := dbanon.NewProvider()
	eav := dbanon.NewEav(config)
	processor := dbanon.NewLineProcessor(mode, config, provider, eav)

	for {
		text, err := reader.ReadString('\n')
		result := processor.ProcessLine(text)
		if mode == "anonymize" {
			fmt.Print(result)
		}

		if err != nil {
			break
		}
	}

	if mode == "map-eav" {
		out, _ := eav.Config.String()
		fmt.Print(string(out))
		os.Exit(0)
	}
}


src/config.go
package dbanon

import (
	"errors"
	"gopkg.in/yaml.v2"
	"io/ioutil"
	"strings"
)

type Config struct {
	Tables []struct {
		Name    string            `yaml:"name"`
		Columns map[string]string `yaml:"columns"`
	}
	Eav []struct {
		Name       string            `yaml:"name"`
		Attributes map[string]string `yaml:"attributes"`
	}
}

func NewConfig(requested string) (*Config, error) {
	c := &Config{}

	if requested == "" {
		return c, errors.New("You must specify a config")
	}

	var source []byte
	var err error

	switch requested {
	default:
		source, err = ioutil.ReadFile(requested)
		if err != nil {
			return c, err
		}
	}

	yaml.Unmarshal(source, &c)

	return c, nil
}

func (c Config) String() ([]byte, error) {
	return yaml.Marshal(c)
}

func (c Config) ProcessTable(t string) string {
	for _, table := range c.Tables {
		if table.Name == t {
			return "table"
		}
	}

	eavSuffixes := [5]string{"_datetime", "_decimal", "_int", "_text", "_varchar"}

	for _, v := range eavSuffixes {
		if strings.HasSuffix(t, v) {
			parts := strings.Split(t, "_entity")
			for _, e := range c.Eav {
				if e.Name == parts[0] {
					return "eav"
				}
			}
		}
	}

	return ""
}

func (c Config) ProcessColumn(t string, col string) (bool, string) {
	for _, table := range c.Tables {
		if table.Name != t {
			continue
		}

		for k, v := range table.Columns {
			if k == col {
				return true, v
			}
		}
	}

	return false, ""
}

func (c Config) ProcessEav(t string, attributeId string) (bool, string) {
	parts := strings.Split(t, "_entity")
	entity := parts[0]
	for _, e := range c.Eav {
		if e.Name == entity {
			for k, v := range e.Attributes {
				if k == attributeId {
					return true, v
				}
			}
		}
	}

	return false, ""
}


src/processor.go
package dbanon

import (
	"strings"

	"vitess.io/vitess/go/vt/sqlparser"
)

type LineProcessor struct {
	Mode     string
	Config   *Config
	Provider ProviderInterface
	Eav      *Eav
}

func NewLineProcessor(m string, c *Config, p ProviderInterface, e *Eav) *LineProcessor {
	return &LineProcessor{Mode: m, Config: c, Provider: p, Eav: e}
}

func (p LineProcessor) ProcessLine(s string) string {
	i := strings.Index(s, "INSERT")
	if i == 0 {
		return p.processInsert(s)
	}

	findNextTable(s)

	return s
}

func (p LineProcessor) processInsert(s string) string {
	stmt, err := sqlparser.Parse(s)
	if err != nil {
		return s
	}
	insert, ok := stmt.(*sqlparser.Insert)

	// This _shouldn't happen but the statement might not be an Insert
	// For example, it'll be nil if the binary charset introducer is foudn
	// https://github.com/blastrain/vitess-sqlparser/issues/25
	if !ok {
		return s
	}

	t, _ := insert.Table.TableName()
	table := t.Name.String()

	processor := p.Config.ProcessTable(table)

	if processor == "" && p.Mode == "anonymize" {
		return s
	}

	var attributeId string
	var result bool
	var dataType string

	var entityTypeId string

	rows := insert.Rows.(sqlparser.Values)
	for _, vt := range rows {
		for i, e := range vt {
			column := currentTable[i].Name

			if processor == "table" && p.Mode == "anonymize" {
				result, dataType = p.Config.ProcessColumn(table, column)

				if !result {
					continue
				}
			}

			switch v := e.(type) {
			case *sqlparser.Literal:
				if processor == "table" {
					v.Val = p.Provider.Get(dataType, &v.Val)
				} else {
					if column == "attribute_id" {
						attributeId = string(v.Val)
						if p.Mode == "anonymize" {
							result, dataType = p.Config.ProcessEav(table, attributeId)
						}
					}
					if column == "value" && result {
						v.Val = p.Provider.Get(dataType, &v.Val)
					}
					if p.Mode == "map-eav" {
						if column == "entity_type_id" {
							entityTypeId = string(v.Val)
						}
						if column == "entity_type_code" {
							p.Eav.entityMap[string(v.Val)] = entityTypeId
						}
						if column == "attribute_code" {
							for _, eavConfig := range p.Eav.Config.Eav {
								if p.Eav.entityMap[eavConfig.Name] == entityTypeId {
									for eavK, eavV := range eavConfig.Attributes {
										if eavK == string(v.Val) {
											eavConfig.Attributes[attributeId] = eavV
										}
									}
								}
							}
						}
					}
				}
			}
		}
	}

	return sqlparser.String(insert) + ";\n"
}


src/create_table.go
package dbanon

import (
	"strings"

	"vitess.io/vitess/go/vt/sqlparser"
)

type Column struct {
	Name string
	Type string
}

func NewColumn(n string, t string) *Column {
	return &Column{Name: n, Type: t}
}

var nextTable = ""
var currentTable = make([]*Column, 0)

func findNextTable(s string) {
	if len(nextTable) > 0 {
		// TODO: Are we guaranteed this will delimit the end of the CREATE TABLE?
		j := strings.Index(s, "/*!40101")
		if j == 0 {
			stmt, _ := sqlparser.Parse(nextTable)
			currentTable = nil
			createTable := stmt.(*sqlparser.CreateTable)
			for _, col := range createTable.TableSpec.Columns {
				column := NewColumn(col.Name.String(), col.Type.Type)
				currentTable = append(currentTable, column)
			}
			nextTable = ""
		} else {
			nextTable += s
		}
	}

	k := strings.Index(s, "CREATE TABLE")
	if k == 0 {
		nextTable += s
	}
}


src/provider.go
package dbanon

import (
	"regexp"
	"strings"
	"math/rand"
	"syreclabs.com/go/faker"
	"github.com/google/uuid"
)

type Provider struct {}

func NewProvider() *Provider {
	p := &Provider{}

	return p
}

type ProviderInterface interface {
	Get(fakeType string, currentValue *string) string
}

func (p Provider) Get(fakeType string, currentValue *string) string {
	if (strings.HasPrefix(fakeType, "dynamic")) {
		return p.processDynamicFake(fakeType, currentValue)
	}

	switch fakeType {
		// Name
		case "first_name":
			return faker.Name().FirstName()
		case "last_name":
			return faker.Name().LastName()
		case "full_name":
			return faker.Name().FirstName() + " " + faker.Name().LastName()
		
		// Company
		case "company_name":
			return faker.Company().Name()

		// Internet
		case "username":
			return faker.Internet().UserName()
		case "password":
			return faker.Internet().Password(8, 14)
		case "ipv4":
			return faker.Internet().IpV4Address()
		case "url":
			return faker.Internet().Url()
		case "linkedin_url":
			return "https://linkedin.com/in/" + faker.Internet().UserName()
		case "md5":
			return faker.Lorem().Characters(32)
		case "uuid":
			return uuid.NewString()
		case "json":
			return p.fakeJson()
		case "query_params":
			return p.fakeQueryParams()

		// Dates
		case "datetime":
			return faker.Date().Birthday(0, 40).Format("2006-01-02 15:04:05")

		// Geo
		case "state":
			return faker.Address().State()
		case "city":
			return faker.Address().City()
		case "postcode":
			return faker.Address().Postcode()
		case "street":
			return faker.Address().StreetAddress()
		case "country_code":
			return faker.Address().CountryCode()

		// Currency
		case "money":
			numberOfDigits := rand.Intn(5 - 1) + 1
			return faker.Number().Number(numberOfDigits)
		case "money_decimal":
			numberOfDigits := rand.Intn(5 - 1) + 1
			return faker.Number().Decimal(numberOfDigits, 2)

		// Contact
		case "email":
			return faker.Internet().Email()
		case "telephone":
			return faker.PhoneNumber().PhoneNumber()
	}

	logger := GetLogger()
	logger.Error(fakeType + " does not match any known type")

	return ""
}

func (p Provider) fakeJson() string {
	return "{\"fake\": true}"
}

func (p Provider) fakeQueryParams() string {
	return "?fake=true"
}

func (p Provider) processDynamicFake(fakeType string, currentValue *string) string {
	// strip dynamic from fakeType
	fakeType = strings.ReplaceAll(fakeType, "dynamic.", "")

	var args []string
	args, fakeType = p.processDynamicArgs(fakeType)

	switch fakeType {
	case "email":
		return p.dynamicEmail(currentValue, args)
	}

	logger := GetLogger()
	logger.Error(fakeType + " does not match any known dynamic type")

	return ""
}

func (p Provider) processDynamicArgs(fakeType string) ([]string, string)  {
	reg := regexp.MustCompile(`\((.*?)\)`)
	args := string(reg.Find([]byte(fakeType)))

	fakeType = strings.ReplaceAll(fakeType, args, "")

	args = strings.TrimPrefix(args, "(")
	args = strings.TrimSuffix(args, ")")

	return strings.Split(args, ","), fakeType
}

func (p Provider) dynamicEmail(currentValue *string, args []string) string {
	if len(args) < 1 || args[0] == "" {
		return faker.Internet().Email()
	}

	// args[0] is email we wish to not fake
	ignoreEmail := args[0]
	val := *currentValue
	if strings.HasSuffix(val, ignoreEmail) {
		return val
	}

	return faker.Internet().Email()
}